lib/spack/spack/build_environment.py

```diff
@@ -452,6 +451,40 @@ def set_build_environment_variables(pkg, env, dirty):
     return env
 
 
+def determine_number_of_jobs(
+        parallel=False, command_line=None, config_default=None, max_cpus=None):
+    """
+    Packages that require sequential builds need 1 job. Otherwise we use the
+    number of jobs set on the command line. If not set, then we use the config
+    defaults (which is usually set through the builtin config scope), but we
+    cap to the number of CPUs available to avoid oversubscription.
+
+    Parameters:
+        parallel (bool): true when package supports parallel builds
+        command_line (int/None): command line override
+        config_default (int/None): config default number of jobs
+        max_cpus (int/None): maximum number of CPUs available. When None, this
+                             value is automatically determined.
+    """
+    if not parallel:
+        return 1
+
+    if command_line is None and 'command_line' in spack.config.scopes():
+        command_line = spack.config.get('config:build_jobs', scope='command_line')
+
+    if command_line is not None:
+        return command_line
+
+    max_cpus = max_cpus or cpus_available()
+
+    config_default = config_default or spack.config.get('config:build_jobs')
+
+    if config_default is None:
+        return max_cpus
```

Comment:
The previous behavior was that 16 was both the config default and the code default if no config was specified. We chose this because someone ran a study that showed there is virtually no performance gains beyond 16 threads for most build jobs.

I like that this allows the user to override that default, but I think we should preserve the code default of 16 if no config specifies a value.

```suggestion
    config_default = config_default or spack.config.get('config:build_jobs', 16)
```

Reply:
I didn't want to duplicate that, since it is already fixed in `spack.config.default_config` which is used for the _builtin scope:

```
$ spack python -c "spack.config.scopes()['_builtin'].get_section('config')['config']['build_jobs']"
16
```

This scope is used almost everywhere, except `when spack.config.use_config(...):` is used and spack.config.default_config is not added as the "bottom" scope (which is the case in tests with config_scope fixtures, and it was the case too during bootstrap, but that's a potential bug addressed in #22610).

Reply:
Oh well, I can just do `, 16) and drop the two lines below that then.

Reply:
done...